README.rst
We'll write a decent readme later

CHANGELOG.txt
Changelog
=========

<<<<<<< HEAD
=======
1.1 (xx.xx.xxxx)
~~~~~~~~~~~~~~~~

 * Implemented the `specific()` method on PageQuerySet, to return pages as their most specific type
 * Implemented pagination in the page chooser modal
 * Changed INSTALLED_APPS in project template to list apps in precedence order (Piet Delport)
 * The `{% image %}` tag now supports filters on the image variable, e.g. `{% image primary_img|default:secondary_img width-500 %}`
 * Moved the style guide menu item into the Settings sub-menu
 * Search backends can now be specified by module (e.g. `wagtail.wagtailsearch.backends.elasticsearch`), rather than a specific class (`wagtail.wagtailsearch.backends.elasticsearch.ElasticSearch`)
 * Added ``descendant_of`` filter to the API (Michael Fillier)
 * Added optional directory argument to "wagtail start" command (Mitchel Cabuloy)
 * Image file size is now stored in the database, to avoid unnecessary filesystem lookups
 * Updated URLs within the admin backend to use namespaces
 * Fix: Text areas in the non-default tab of the page editor now resize to the correct height
 * Fix: Tabs in "insert link" modal in the rich text editor no longer disappear (Tim Heap)
 * Fix: H2 elements in rich text fields were accidentally given a click() binding when put insite a collapsible multi field panel


>>>>>>> d4259e133b80d5648266db181029dfbe0fbcf885
1.0 (16.07.2015)
~~~~~~~~~~~~~~~~

 * Added StreamField, a model field for freeform page content
 * Added Wagtail API, a module for creating a RESTful API for your content
 * MySQL support
 * Django 1.8 support
 * Removed dependency on libsass (Tim Heap)
 * Users without usernames can now be created and edited in the admin interface (Tim Heap)
 * Added update notifications
 * Javascript includes in the admin backend have been moved to the HTML header, to accommodate form widgets that render inline scripts that depend on libraries such as jQuery
 * Improvements to the layout of the admin menu footer.
 * Menu items of custom apps are now highlighted when being used (Josh Barr)
 * Added thousands separator for counters on dashboard
 * Added contextual links to admin notification messages
 * When copying pages, it is now possible to specify a place to copy to (Timo Rieber)
 * FieldPanel now accepts an optional 'widget' parameter to override the field's default form widget (Alejandro Giacometti)
 * Page URL paths can now be longer than 255 characters
 * Dropped Django 1.6 support
 * Dropped Python 2.6 and 3.2 support
 * Dropped Elasticsearch 0.90.x support
 * Serving documents will now use django-sendfile if it's configured (Jordi Joan)
 * Documents are now served with correct mime-type (Jordi Joan, Damian Moore)
 * Support for If-Modified-Since HTTP header (Jordi Joan)
 * Search view accepts "page" GET parameter in line with pagination
 * Reversing `django.contrib.auth.admin.login` will no longer lead to Wagtails login view (making it easier to have front end views)
 * Removed dependency on `LOGIN_URL` and `LOGIN_REDIRECT_URL` settings
 * Password reset view names namespaced to wagtailadmin
 * Removed the need to add permission check on admin views (now automated)
 * Added cache-control headers to all admin views
 * Page model fields without a FieldPanel are no longer displayed in the form
 * No longer need to specify the base model on InlinePanel definitions
 * The project template Vagrantfile now listens on port 8000
 * The external link chooser in rich text areas now accepts URLs of the form '/some/local/path', to allow linking to non-Wagtail-controlled URLs within the local site (Eric Drechsel)
 * SCSS files in wagtailadmin now use absolute imports, to permit overriding by user stylesheets (Martin Sanders)
 * Bare text entered in rich text areas is now automatically wrapped in a paragraph element
 * Added pagination to the snippets listing and chooser (Martin Sanders)
 * Page / document / image / snippet choosers now include a link to edit the chosen item
 * The `document_served` signal now correctly passes the Document class as `sender` and the document as `instance`
 * Image/Document edit page no longer throws OSError when the original image is missing
 * Page classes can specify an edit_handler property to override the default Content / Promote / Settings tabbed interface
 * The Page model now records the date/time that a page was first published, as the field `first_published_at`
 * Increased the maximum length of a page slug from 50 to 255 characters
 * Plain text fields in the page editor now use auto-expanding text areas
 * Date / time pickers now consistently use times without seconds, to prevent Javascript behaviour glitches when focusing / unfocusing fields
 * Added hooks `register_rich_text_embed_handler` and `register_rich_text_link_handler` for customising link / embed handling within rich text fields
 * Added hook `construct_homepage_summary_items` for customising the site summary panel on the admin homepage
 * No longer automatically tries to use Celery for sending notification emails
 * Added "Add child page" button to admin userbar (Eric Drechsel)
 * Renamed the `construct_wagtail_edit_bird` hook to `construct_wagtail_userbar`
 * 'static' template tags are now used throughout the admin templates, in place of STATIC_URL
 * Added a new decorator-based syntax for RoutablePage, compatible with Django 1.8
 * Collapsible blocks stay open on any form error (Salvador Faria)
 * Document upload modal no longer switches tabs on form errors (Salvador Faria)
 * Added `AUTO_UPDATE` flag to search backend settings to enable/disable automatically updating the search index on model changes
 * Made the built-in project template follow the Django one, with several Wagtail-specific additions. The template comes with two apps (home and search)
 * `with_metaclass` is now imported from Django's bundled copy of the `six` library, to avoid errors on Mac OS X from an outdated system copy of the library being imported

wagtail/wagtailcore/utils.py
import re
import unicodedata
<<<<<<< HEAD
from six import string_types
=======
>>>>>>> d4259e133b80d5648266db181029dfbe0fbcf885

from django.db.models import Model
from django.apps import apps
from django.utils.encoding import force_text
from django.utils.text import slugify
<<<<<<< HEAD
=======
from django.utils.six import string_types
>>>>>>> d4259e133b80d5648266db181029dfbe0fbcf885


def camelcase_to_underscore(str):
    # http://djangosnippets.org/snippets/585/
    return re.sub('(((?<=[a-z])[A-Z])|([A-Z](?![A-Z]|$)))', '_\\1', str).lower().strip('_')


def resolve_model_string(model_string, default_app=None):
    """
    Resolve an 'app_label.model_name' string into an actual model class.
    If a model class is passed in, just return that.
    """
    if isinstance(model_string, string_types):
        try:
            app_label, model_name = model_string.split(".")
        except ValueError:
            if default_app is not None:
                # If we can't split, assume a model in current app
                app_label = default_app
                model_name = model_string
            else:
                raise ValueError("Can not resolve {0!r} into a model. Model names "
                                 "should be in the form app_label.model_name".format(
                                     model_string), model_string)

        return apps.get_model(app_label, model_name)

    elif isinstance(model_string, type) and issubclass(model_string, Model):
        return model_string

    else:
        raise LookupError("Can not resolve {0!r} into a model".format(model_string), model_string)


SCRIPT_RE = re.compile(r'<(-*)/script>')
def escape_script(text):
    """
    Escape `</script>` tags in 'text' so that it can be placed within a `<script>` block without
    accidentally closing it. A '-' character will be inserted for each time it is escaped:
    `<-/script>`, `<--/script>` etc.
    """
    return SCRIPT_RE.sub(r'<-\1/script>', text)


SLUGIFY_RE = re.compile(r'[^\w\s-]', re.UNICODE)


def cautious_slugify(value):
    """
    Convert a string to ASCII exactly as Django's slugify does, with the exception
    that any non-ASCII alphanumeric characters (that cannot be ASCIIfied under Unicode
    normalisation) are escaped into codes like 'u0421' instead of being deleted entirely.

    This ensures that the result of slugifying e.g. Cyrillic text will not be an empty
    string, and can thus be safely used as an identifier (albeit not a human-readable one).
    """
    value = force_text(value)

    # Normalize the string to decomposed unicode form. This causes accented Latin
    # characters to be split into 'base character' + 'accent modifier'; the latter will
    # be stripped out by the regexp, resulting in an ASCII-clean character that doesn't
    # need to be escaped
    value = unicodedata.normalize('NFKD', value)

    # Strip out characters that aren't letterlike, underscores or hyphens,
    # using the same regexp that slugify uses. This ensures that non-ASCII non-letters
    # (e.g. accent modifiers, fancy punctuation) get stripped rather than escaped
    value = SLUGIFY_RE.sub('', value)

    # Encode as ASCII, escaping non-ASCII characters with backslashreplace, then convert
    # back to a unicode string (which is what slugify expects)
    value = value.encode('ascii', 'backslashreplace').decode('ascii')

    # Pass to slugify to perform final conversion (whitespace stripping, applying
    # mark_safe); this will also strip out the backslashes from the 'backslashreplace'
    # conversion
    return slugify(value)


wagtail/wagtailcore/migrations/0002_initial_data.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django import VERSION as DJANGO_VERSION
from django.db import migrations


def initial_data(apps, schema_editor):
    ContentType = apps.get_model('contenttypes.ContentType')
    Group = apps.get_model('auth.Group')
    Page = apps.get_model('wagtailcore.Page')
    Site = apps.get_model('wagtailcore.Site')
    GroupPagePermission = apps.get_model('wagtailcore.GroupPagePermission')

    # Create page content type
    page_content_type, created = ContentType.objects.get_or_create(
        model='page',
        app_label='wagtailcore',
        defaults={'name': 'page'} if DJANGO_VERSION < (1, 8) else {}
    )

    # Create root page
    root = Page.objects.create(
        title="Root",
        slug='root',
        content_type=page_content_type,
        path='0001',
        depth=1,
        numchild=1,
        url_path='/',
    )

    # Create homepage
    homepage = Page.objects.create(
        title="Welcome to your new BalkanCMS site!",
        slug='home',
        content_type=page_content_type,
        path='00010001',
        depth=2,
        numchild=0,
        url_path='/home/',
    )

    # Create default site
    Site.objects.create(
        hostname='localhost',
        root_page_id=homepage.id,
        is_default_site=True
    )

    # Create auth groups
    moderators_group = Group.objects.create(name='Moderators')
    editors_group = Group.objects.create(name='Editors')

    # Create group permissions
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='add',
    )
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='edit',
    )
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='publish',
    )

    GroupPagePermission.objects.create(
        group=editors_group,
        page=root,
        permission_type='add',
    )
    GroupPagePermission.objects.create(
        group=editors_group,
        page=root,
        permission_type='edit',
    )


class Migration(migrations.Migration):

    dependencies = [
        ('wagtailcore', '0001_initial'),
    ]

    operations = [
        migrations.RunPython(initial_data),
    ]


wagtail/wagtailcore/migrations/0001_squashed_0016_change_page_url_path_to_text_field.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import models, migrations
from django.conf import settings
import django.db.models.deletion
from django import VERSION as DJANGO_VERSION
import wagtail.wagtailsearch.index


def initial_data(apps, schema_editor):
    ContentType = apps.get_model('contenttypes.ContentType')
    Group = apps.get_model('auth.Group')
    Page = apps.get_model('wagtailcore.Page')
    Site = apps.get_model('wagtailcore.Site')
    GroupPagePermission = apps.get_model('wagtailcore.GroupPagePermission')

    # Create page content type
    page_content_type, created = ContentType.objects.get_or_create(
        model='page',
        app_label='wagtailcore',
        defaults={'name': 'page'} if DJANGO_VERSION < (1, 8) else {}
    )

    # Create root page
    root = Page.objects.create(
        title="Root",
        slug='root',
        content_type=page_content_type,
        path='0001',
        depth=1,
        numchild=1,
        url_path='/',
    )

    # Create homepage
    homepage = Page.objects.create(
        title="Welcome to your new BalkanCMS site!",
        slug='home',
        content_type=page_content_type,
        path='00010001',
        depth=2,
        numchild=0,
        url_path='/home/',
    )

    # Create default site
    Site.objects.create(
        hostname='localhost',
        root_page_id=homepage.id,
        is_default_site=True
    )

    # Create auth groups
    moderators_group = Group.objects.create(name='Moderators')
    editors_group = Group.objects.create(name='Editors')

    # Create group permissions
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='add',
    )
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='edit',
    )
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='publish',
    )

    GroupPagePermission.objects.create(
        group=editors_group,
        page=root,
        permission_type='add',
    )
    GroupPagePermission.objects.create(
        group=editors_group,
        page=root,
        permission_type='edit',
    )


    # 0005 - add_page_lock_permission_to_moderators
    GroupPagePermission.objects.create(
        group=moderators_group,
        page=root,
        permission_type='lock',
    )


class Migration(migrations.Migration):

    replaces = [
        ('wagtailcore', '0001_initial'),
        ('wagtailcore', '0002_initial_data'),
        ('wagtailcore', '0003_add_uniqueness_constraint_on_group_page_permission'),
        ('wagtailcore', '0004_page_locked'),
        ('wagtailcore', '0005_add_page_lock_permission_to_moderators'),
        ('wagtailcore', '0006_add_lock_page_permission'),
        ('wagtailcore', '0007_page_latest_revision_created_at'),
        ('wagtailcore', '0008_populate_latest_revision_created_at'),
        ('wagtailcore', '0009_remove_auto_now_add_from_pagerevision_created_at'),
        ('wagtailcore', '0010_change_page_owner_to_null_on_delete'),
        ('wagtailcore', '0011_page_first_published_at'),
        ('wagtailcore', '0012_extend_page_slug_field'),
        ('wagtailcore', '0013_update_golive_expire_help_text'),
        ('wagtailcore', '0014_add_verbose_name'),
        ('wagtailcore', '0015_add_more_verbose_names'),
        ('wagtailcore', '0016_change_page_url_path_to_text_field')
    ]

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
        ('auth', '0001_initial'),
        ('contenttypes', '0001_initial'),
    ]

    operations = [
        migrations.CreateModel(
            name='Page',
            fields=[
                ('id', models.AutoField(auto_created=True, verbose_name='ID', serialize=False, primary_key=True)),
                ('path', models.CharField(unique=True, max_length=255)),
                ('depth', models.PositiveIntegerField()),
                ('numchild', models.PositiveIntegerField(default=0)),
                ('title', models.CharField(verbose_name='Title', max_length=255, help_text="The page title as you'd like it to be seen by the public")),
                ('slug', models.SlugField(verbose_name='Slug', max_length=255, help_text='The name of the page as it will appear in URLs e.g http://domain.com/blog/[my-slug]/')),
                ('live', models.BooleanField(default=True, verbose_name='Live', editable=False)),
                ('has_unpublished_changes', models.BooleanField(default=False, verbose_name='Has unpublished changes', editable=False)),
                ('url_path', models.TextField(verbose_name='URL path', blank=True, editable=False)),
                ('seo_title', models.CharField(verbose_name='Page title', max_length=255, blank=True, help_text="Optional. 'Search Engine Friendly' title. This will appear at the top of the browser window.")),
                ('show_in_menus', models.BooleanField(default=False, verbose_name='Show in menus', help_text='Whether a link to this page will appear in automatically generated menus')),
                ('search_description', models.TextField(verbose_name='Search description', blank=True)),
                ('go_live_at', models.DateTimeField(null=True, verbose_name='Go live date/time', blank=True, help_text='Please add a date-time in the form YYYY-MM-DD hh:mm.')),
                ('expire_at', models.DateTimeField(null=True, verbose_name='Expiry date/time', blank=True, help_text='Please add a date-time in the form YYYY-MM-DD hh:mm.')),
                ('expired', models.BooleanField(default=False, verbose_name='Expired', editable=False)),
                ('content_type', models.ForeignKey(verbose_name='Content type', related_name='pages', to='contenttypes.ContentType')),
                ('owner', models.ForeignKey(null=True, verbose_name='Owner', blank=True, editable=False, related_name='owned_pages', to=settings.AUTH_USER_MODEL, on_delete=django.db.models.deletion.SET_NULL)),
                ('locked', models.BooleanField(default=False, verbose_name='Locked', editable=False)),
                ('latest_revision_created_at', models.DateTimeField(null=True, verbose_name='Latest revision created at', editable=False)),
                ('first_published_at', models.DateTimeField(null=True, verbose_name='First published at', editable=False)),
            ],
            options={
                'abstract': False,
            },
            bases=(models.Model, wagtail.wagtailsearch.index.Indexed),
        ),
        migrations.CreateModel(
            name='GroupPagePermission',
            fields=[
                ('id', models.AutoField(auto_created=True, verbose_name='ID', serialize=False, primary_key=True)),
                ('permission_type', models.CharField(verbose_name='Permission type', choices=[('add', 'Add/edit pages you own'), ('edit', 'Add/edit any page'), ('publish', 'Publish any page'), ('lock', 'Lock/unlock any page')], max_length=20)),
                ('group', models.ForeignKey(verbose_name='Group', related_name='page_permissions', to='auth.Group')),
                ('page', models.ForeignKey(verbose_name='Page', related_name='group_permissions', to='wagtailcore.Page')),
            ],
        ),
        migrations.AlterUniqueTogether(
            name='grouppagepermission',
            unique_together=set([('group', 'page', 'permission_type')]),
        ),
        migrations.AlterModelOptions(
            name='grouppagepermission',
            options={'verbose_name': 'Group Page Permission'},
        ),
        migrations.CreateModel(
            name='PageRevision',
            fields=[
                ('id', models.AutoField(auto_created=True, verbose_name='ID', serialize=False, primary_key=True)),
                ('submitted_for_moderation', models.BooleanField(default=False, verbose_name='Submitted for moderation')),
                ('created_at', models.DateTimeField(verbose_name='Created at')),
                ('content_json', models.TextField(verbose_name='Content JSON')),
                ('approved_go_live_at', models.DateTimeField(null=True, verbose_name='Approved go live at', blank=True)),
                ('page', models.ForeignKey(verbose_name='Page', related_name='revisions', to='wagtailcore.Page')),
                ('user', models.ForeignKey(null=True, verbose_name='User', blank=True, to=settings.AUTH_USER_MODEL)),
            ],
        ),
        migrations.AlterModelOptions(
            name='pagerevision',
            options={'verbose_name': 'Page Revision'},
        ),
        migrations.CreateModel(
            name='PageViewRestriction',
            fields=[
                ('id', models.AutoField(auto_created=True, verbose_name='ID', serialize=False, primary_key=True)),
                ('password', models.CharField(verbose_name='Password', max_length=255)),
                ('page', models.ForeignKey(verbose_name='Page', related_name='view_restrictions', to='wagtailcore.Page')),
            ],
        ),
        migrations.AlterModelOptions(
            name='pageviewrestriction',
            options={'verbose_name': 'Page View Restriction'},
        ),
        migrations.CreateModel(
            name='Site',
            fields=[
                ('id', models.AutoField(auto_created=True, verbose_name='ID', serialize=False, primary_key=True)),
                ('hostname', models.CharField(verbose_name='Hostname', db_index=True, max_length=255)),
                ('port', models.IntegerField(default=80, verbose_name='Port', help_text='Set this to something other than 80 if you need a specific port number to appear in URLs (e.g. development on port 8000). Does not affect request handling (so port forwarding still works).')),
                ('is_default_site', models.BooleanField(default=False, verbose_name='Is default site', help_text='If true, this site will handle requests for all other hostnames that do not have a site entry of their own')),
                ('root_page', models.ForeignKey(verbose_name='Root page', related_name='sites_rooted_here', to='wagtailcore.Page')),
            ],
        ),
        migrations.AlterUniqueTogether(
            name='site',
            unique_together=set([('hostname', 'port')]),
        ),
        migrations.AlterModelOptions(
            name='site',
            options={'verbose_name': 'Site'},
        ),
        migrations.RunPython(
            initial_data,
        ),
    ]
